### PR TITLE
Skills cleanup and the fallout.

### PR DESCRIPTION
**Pull request by bkentel**
The rabbit hole was deeper than expected: lots of side-changes (most are minor touch ups though). I originally planned to touch only `time_to_fire`, but then something wasn't `const`, and then I noticed an unneeded `new` /`delete` and cleaned up the fallout...

Most of the actual changes are pretty straight forward, but due to the size of this some extra eyes on the code would be for the best.
- Added a few overloads that work for `Skill const&`.
- Applied `const` where appropriate in places touched by this PR.
- Added `Skill::get_skills_sorted_by` to encapsulate the common logic needed to get a sorted vector.
- Added `SkillLevel::exercised_level` to centralize some duplicate code.
- Removed the explicit new / delete in favor of a vector of values.
- Replace a few for loops with range-based for loops.
- Remove one-off algorithm helper functions and flip them to lambdas.
- Collected the repeated logic in `time_to_fire` into a static map.
- Cleaned up `Skill`'s constructors to use initializers and delegating constructors.
- Hopefully clarified `rustRate`.

### CODE REVIEW

**Comment by kevingranade** [comment]
1. Not even looking at this until after 0.C
2. Based on your description there's a good chance I'm going to refuse to
merge it without you breaking it up, it sounds like there are a lot of
unrelated things merged together in here.

**Comment by bkentel** [comment]
(1) That's fine.
(2) Not really. Maybe one portion could be made a separate PR the rest is rather inextricably tied together. Though some of it could be done piecemeal. I'll see if I can can chop it up into more manageable pieces when I have some time.

**Comment by kevingranade** [comment]
I did a mega-merge, and after building was greeted with this backtrace:
I might be jumping to a conclusion, but this PR rewrote the indicated file so it seems likely.
```
Core was generated by `./cataclysm'.
Program terminated with signal SIGABRT, Aborted.
#0  0x00007f09f19ec4b7 in raise () from /usr/lib/libc.so.6
(gdb) bt
#0  0x00007f09f19ec4b7 in raise () from /usr/lib/libc.so.6
#1  0x00007f09f19ed88a in abort () from /usr/lib/libc.so.6
#2  0x00007f09f232f815 in __gnu_debug::_Error_formatter::_M_error() const () from /usr/lib/libstdc++.so.6
#3  0x0000000000d1e501 in __gnu_debug::_Safe_iterator<__gnu_cxx::__normal_iterator<std::pair<Skill const*, int>*, std::__cxx1998::vector<std::pair<Skill const*, int>, std::allocator<std::pair<Skill const*, int> > > >, std::__debug::vector<std::pair<Skill const*, int>, std::allocator<std::pair<Skill const*, int> > > >::operator-> (this=0x7fff31380a70) at /usr/include/c++/4.9.2/debug/safe_iterator.h:273
#4  0x0000000000d11cc9 in npc::best_skill (this=0x3d0b190) at src/npc.cpp:965
#5  0x0000000000d171a7 in npc::value (this=0x3d0b190, it=...) at src/npc.cpp:1586
#6  0x00000000005b15cd in npc::find_item (this=0x3d0b190) at src/npcmove.cpp:1342
#7  0x00000000005a85af in npc::move (this=0x3d0b190) at src/npcmove.cpp:161
#8  0x0000000000a0037f in game::monmove (this=0x7f09f1955010) at src/game.cpp:6006
#9  0x00000000009ded8e in game::do_turn (this=0x7f09f1955010) at src/game.cpp:1485
#10 0x0000000000a95421 in main (argc=0, argv=0x7fff31381f30) at src/main.cpp:434
```

**Comment by bkentel** [comment]
@kevingranade How was this failure triggered? This happened with the current master branch + this PR?

**Comment by kevingranade** [comment]
There were some other PRs involved, but they didn't have anything to do
with this piece of code.  I just wandered in range of a NPC and got the
segfault.  I'll see if I can reproduce with just this one later.

**Comment by bkentel** [comment]
Thanks, I'll do some testing of my own in the meantime too to see if I can reproduce it and fix whatever's causing it.

**Comment by bkentel** [comment]
Looks like I can trigger it with an NPC too; working on a fix.

**Comment by narc0tiq** [comment]
Something went wrong: http://ci.narc.ro/job/Cataclysm-Matrix/2868/

File: src/skill.cpp [hunk 1]
```diff
@@ -160,23 +148,27 @@ bool SkillLevel::isRusting() const
...
+        if (rust_type == "vanilla" || rust_type == "int") {
+            _exercise = (100 * _level) - 1;
+            --_level;
+        } else {
+            _exercise = 0;
         }
     }
-    return false;
+
+    return true;
```
Comment (BevapDin) [inline src/skill.cpp]: This changes the behavior of the function. If it returns true, the characters bionic power will be reduced. This causes #11623, see `game::rustCheck`. There is no additional check for the memory bionic being present, this function should *only* return true if `charged_bio_mem` is true. In turn, that parameter will only be true if the character has the memory bionic.

Comment (bkentel) [inline src/skill.cpp]: :frowning: Not sure how false became true, but I'll make a PR if one hasn't already been made; thanks @BevapDin .

**Pull request merged**